File: sdk/arena-java-sdk/src/main/java/com/github/kubeflow/arena/client/ServingClient.java
package com.github.kubeflow.arena.client;

import com.alibaba.fastjson.JSONObject;
import com.github.kubeflow.arena.enums.ArenaErrorEnum;
import com.github.kubeflow.arena.enums.ServingJobType;
import com.github.kubeflow.arena.exceptions.ArenaException;
import com.github.kubeflow.arena.exceptions.ExitCodeException;
import com.github.kubeflow.arena.model.serving.ServingJob;
import com.github.kubeflow.arena.model.serving.ServingJobInfo;
import com.github.kubeflow.arena.utils.Command;

import java.io.IOException;
import java.util.List;

public class ServingClient extends BaseClient {

    public ServingClient(String kubeConfig, String namespace, String loglevel, String arenaSystemNamespace) {
        super(kubeConfig, namespace, loglevel, arenaSystemNamespace);
    }

    public ServingClient namespace(String namespace) {
        return new ServingClient(this.kubeConfig, namespace, this.loglevel, this.arenaSystemNamespace);
    }

    public String submit(ServingJob job) throws ArenaException, IOException {
        List<String> cmds = this.generateCommands("serve", job.getType().shortHand());
        for (int i = 0; i < job.getArgs().size(); i++) {
            cmds.add(job.getArgs().get(i));
        }
        cmds.add(job.getCommand());
        String[] arenaCommand = cmds.toArray(new String[cmds.size()]);
        try {
            String output = Command.execCommand(arenaCommand);
            return output;
        } catch (ExitCodeException e) {
            if (e.getMessage().contains(String.format("the job %s is already exist, please delete it first.", job.name()))) {
                throw new ArenaException(ArenaErrorEnum.SERVING_JOB_EXISTS, e.getMessage());
            } else {
                throw new ArenaException(ArenaErrorEnum.SERVING_SUBMIT, e.getMessage());
            }
        }
    }

    public List<ServingJobInfo> list(ServingJobType jobType) throws ArenaException, IOException {
        return this.list(jobType, null);
    }

    public List<ServingJobInfo> list(ServingJobType jobType, Boolean allNamespaces) throws ArenaException, IOException {
        List<String> cmds = this.generateCommands("serve", "list");
        if (!jobType.equals(ServingJobType.AllServingJob) && !jobType.equals(ServingJobType.UnknownServingJob)) {
            cmds.add("--type=" + jobType.shortHand());
        }

        if (allNamespaces != null && allNamespaces) {
            cmds.add("-A");
        }

        cmds.add("-o");
        cmds.add("json");
        String[] arenaCommand = cmds.toArray(new String[cmds.size()]);
        try {
            String output = Command.execCommand(arenaCommand);
            return JSONObject.parseArray(output, ServingJobInfo.class);
        } catch (ExitCodeException e) {
            throw new ArenaException(ArenaErrorEnum.SERVING_LIST, e.getMessage());
        }
    }

    public ServingJobInfo get(String jobName, ServingJobType jobType, String jobVersion) throws ArenaException, IOException {
        List<String> cmds = this.generateCommands("serve", "get");
        if (!jobType.equals(ServingJobType.AllServingJob) && !jobType.equals(ServingJobType.UnknownServingJob)) {
            cmds.add("--type=" + jobType.shortHand());
        }
        if (jobVersion != null && jobVersion.length() != 0) {
            cmds.add("--version=" + jobVersion);
        }
        cmds.add(jobName);
        cmds.add("-o");
        cmds.add("json");
        String[] arenaCommand = cmds.toArray(new String[cmds.size()]);
        ServingJobInfo jobInfo = new ServingJobInfo();
        try {
            String output = Command.execCommand(arenaCommand);
            jobInfo = jobInfo.complete(output);
            return jobInfo;
        } catch (ExitCodeException e) {
            if (e.getMessage().contains(String.format("Not found serving job %s, please check it with `arena serve list | grep %s`", jobName, jobName))) {
                return null;
            } else {
                throw new ArenaException(ArenaErrorEnum.SERVING_GET, e.getMessage());
            }
        }
    }

    public String delete(String jobName, ServingJobType jobType, String jobVersion) throws IOException, ArenaException {
        List<String> cmds = this.generateCommands("serve", "delete");
        if (!jobType.equals(ServingJobType.AllServingJob) && !jobType.equals(ServingJobType.UnknownServingJob)) {
            cmds.add("--type=" + jobType.shortHand());
        }
        if (jobVersion != null && jobVersion.length() != 0) {
            cmds.add("--version=" + jobVersion);
        }
        cmds.add(jobName);
        String[] arenaCommand = cmds.toArray(new String[cmds.size()]);
        try {
            String output = Command.execCommand(arenaCommand);
            return output;
        } catch (ExitCodeException e) {
            throw new ArenaException(ArenaErrorEnum.SERVING_DELETE, e.getMessage());
        }
    }

}

File: sdk/arena-python-sdk/arenasdk/client/serving_client.py
#!/usr/bin/env python
from __future__ import annotations
import json
from typing import List
from typing import Dict
from arenasdk.enums.types import *
from arenasdk.serving.job import ServingJob
from arenasdk.serving.serving_job_info import ServingJobInfo
from arenasdk.serving.serving_job_info import generate_serving_job_info
from arenasdk.common.util import Command
from arenasdk.exceptions.arena_exception import ArenaException

class ServingClient(object):
    def __init__(self,kubeconfig: str,namespace: str,loglevel: str,arena_namespace: str):
        self.kubeconfig = kubeconfig
        self.namespace = namespace
        self.loglevel = loglevel
        self.arena_namespace = arena_namespace

    def namespace(self,namespace: str) -> TrainingClient:
        return ServingClient(self.kubeconfig,namespace,self.loglevel,self.arena_namespace)

    def submit(self,job: TrainingJob) -> str:
        cmds = self.__generate_commands("serve")
        cmds.append(job.get_type().value[0])
        for opt in job.get_args():
            cmds.append(opt)
        cmds.append(job.get_command())
        try:
            status,stdout,stderr = Command(*cmds).run()
            if status != 0:
                err_message = "the job {} is already exist, please delete it first.".format(job.get_name())
                if stdout.find(err_message) != -1 or stderr.find(err_message) != -1:
                    raise ArenaException(ArenaErrorType.TrainingJobExistError,stdout + stderr)
                else:
                    raise ArenaException(ArenaErrorType.SubmitTrainingJobError,stdout + stderr)
            return stdout
        except ArenaException as e:
            raise e
    
    def list(self,job_type: ServingJobType,all_namespaces: bool) -> List[TrainingJobInfo]:
        def convert(json_object):
            job_infos = list()
            for obj in json_object:
                job_infos.append(generate_serving_job_info(obj))
            return job_infos
        cmds = self.__generate_commands("list")
        if job_type != ServingJobType.AllServingJob and job_type != ServingJobType.UnknownServingJob:
            cmds.append("--type=" + job_type.value)
        if all_namespaces:
            cmds.append("-A")
        cmds.append("-o")
        cmds.append("json")
        try:
            status,stdout,stderr = Command(*cmds).run()
            if status != 0:
                raise ArenaException(ArenaErrorType.ListServingJobError,stdout + stderr)
            return json.loads(stdout,object_hook=convert)
        except ArenaException as e:
            raise e

    def get(self,job_name: str,job_type: ServingJobType,version: str) -> TrainingJobInfo:
        cmds = self.__generate_commands("serve","get")
        cmds.append(job_name)
        if job_type != ServingJobType.AllServingJob and job_type != ServingJobType.UnknownServingJob:
            cmds.append("--type=" + job_type.value[0])
        if version != "":
            cmds.append("--version=" + version)
        cmds.append("-o")
        cmds.append("json")
        try:
            status,stdout,stderr = Command(*cmds).run()
            if status != 0:
                err_message = "Not found serving job {}".format(job_name)
                if stdout.find(err_message) != -1 or stderr.find(err_message) != -1:
                    return None
                else:
                    raise ArenaException(ArenaErrorType.GetServingJobError,stdout + stderr)
            data = json.loads(stdout)
            return generate_serving_job_info(data)
        except ArenaException as e:
            raise e
    
    def delete(self, job_name: str,job_type: ServingJobType,version: str) -> str:
        cmds = self.__generate_commands("serve","delete")
        cmds.append(job_name)
        if job_type != TrainingJobType.AllTrainingJob and job_type != TrainingJobType.UnknownTrainingJob:
            cmds.append("--type=" + job_type.value[0])
        if version != "":
            cmds.append("--version=" + version)
        try:
            status,stdout,stderr = Command(*cmds).run()
            if status != 0:
                raise ArenaException(ArenaErrorType.DeleteServingJobError,stdout + stderr)
            return stdout
        except ArenaException as e:
            raise e

    def traffic_router_split(self,job_name: str,job_type: ServingJobType,version_weights: Dict[str,int]):
        cmds = self.__generate_commands("serve","traffic-router-split")
        cmds.append("--name=" + job_name)
        for version,weight in version_weights.items():
            cmds.append("-v=" + version + ":" + str(weight))
        try:
            status,stdout,stderr = Command(*cmds).run()
            if status != 0:
                raise ArenaException(ArenaErrorType.TrafficRouterSplitServingJobError,stdout + stderr)
            return stdout
        except ArenaException as e:
            raise e
        
         
    def __generate_commands(self,*sub_commands: List[str]) -> List[str]:
        arena_cmds = list()
        arena_cmds.append(ARENA_BINARY)
        for c in sub_commands:
            arena_cmds.append(c)
        if self.kubeconfig != "":
            arena_cmds.append("--config=" + self.kubeconfig)
        if self.namespace != "":
            arena_cmds.append("--namespace=" + self.namespace)
        if self.arena_namespace != "":
            arena_cmds.append("--arena-namespace=" + self.arena_namespace)
        if self.loglevel != "":
            arena_cmds.append("--loglevel=" + self.loglevel)
        return arena_cmds
        		
